peparser.cpp

```diff
@@ -1323,38 +1323,49 @@ namespace peparser
 		return false;
 	}
 
-	bool PEParser::DetectMIDLMarker(size_t diffStart, size_t diffSize, size_t& diffShift) const
+	bool PEParser::DetectMIDLMarker(size_t diffStart, size_t diffSize, size_t& diffShift, bool tlbCmpExpr) const
 	{
-		diffShift = 0;
+        diffShift = 0;
 
-		UsefulBlockMapRange markers = m_useful.equal_range(MidlTimestampSegment);
-		if(markers.first == markers.second) return false;
+        if (!tlbCmpExpr)
+            return false;
 
-		Block marker;
+        const auto it = std::find_if(cbegin(m_resourceBlocks), cend(m_resourceBlocks), [](const Block& block) {
+            return std::wstring::npos != block.description.find(L"@TYPELIB");
+        });
 
-		for(UsefulBlockMap::const_iterator it = markers.first; it != markers.second; ++it)
-		{
-			if(diffStart < it->second.offset || diffStart >= it->second.offset + it->second.size)
-				continue;
+        if (cend(m_resourceBlocks) != it)
+        {
+            const std::string magic = "Created by MIDL version";
 
-			marker = it->second;
-			break;
-		}
+            const char* stringStart = FindStringEntry< char, None >(
+                (char*)m_view + it->offset, magic, magic.size(), it->size);
 
-		if(marker.offset == 0) 
-			return false;
+            if (stringStart == nullptr)
+                return false;
 
-		const char* sectionStart = (char*)((LPBYTE)m_view + marker.offset);
+            const struct
+            {
+                const unsigned char byte;
+                const size_t offset;
 
-		std::string magic = "Created by MIDL version";
-		const char* stringStart = FindStringEntry<char, None>(sectionStart, magic, magic.size(), marker.size);
+                bool check(const void* data) const noexcept {
+                    return data && byte == *((unsigned char *)data + offset);
+                }
+            }
+            lf{ 0x0a, 61 }, dc3{ 0x13, 62 };
 
-		if(stringStart == NULL) 
-			return false;
+            if (lf.check(stringStart) && dc3.check(stringStart))
+            {
+                const size_t tlbStampLength = 65;
+                const auto diffPos = (char*)m_view + diffStart;
 
-		diffShift = marker.size - diffStart + marker.offset;
+                if (stringStart < diffPos && diffPos < stringStart + tlbStampLength)
+                    return true;
```

Comment:
Yep, size of the ignored block. 

`diffShift` is used to skip both files by the number of bytes (because you detected a part that can be ignored and you know its full size before hand). If both files shift by the same amount, Compare skips.

If you don't shift, and the section has similar bytes in the middle, detection will run for each independent part. For example one binary has ignorable string `AAAAX1111` and second has `BBBBX0000` in the same place, Compare will call heuristics on `AAAA` vs `BBBBB`, one of the checks detects that whole 9 bytes can be ignored and shifts by 9. Compare then restarts after the end. 

If check doesn't shift, Compare will call heuristics for `1111` vs `0000`, because `X` matches in both places and interrupts the difference. The check then needs to know how to detect whatever it is detecting from the middle. (this is what `DetectFILEMacro` does -- it looks back to see if bytes behind look anything like pdb path.)